xdsl/transforms/mpi_to_llvm.py

```diff
@@ -0,0 +1,525 @@
+    }
+    """
+    Translation table for mpi operation names to their MPI library function names
+    """
+
+    info: MpiLibraryInfo
+    """
+    This object carries information about the targeted MPI libraray
+    """
+
+    def __init__(self, info: MpiLibraryInfo):
+        self.info = info
+
+    # Helpers
+
+    def _emit_mpi_status_obj(
+        self, mpi_status_none: bool
+    ) -> tuple[list[Operation], list[SSAValue | None], Operation]:
+        """
+        This function create operations that instantiate a pointer to an MPI_Status-sized object.
+
+        If mpi_status_none = True is passed, it instead loads the magic value MPI_STATUS_IGNORE
+
+        This is currently OpenMPI specific code, as other implementations probably have a different
+        magic value for MPI_STATUS_NONE.
+        """
+        if mpi_status_none:
+            return [
+                lit1 := arith.Constant.from_int_and_width(1, builtin.i64),
+                res := llvm.IntToPtrOp.get(lit1),
+            ], [], res
+        else:
+            return [
+                lit1 := arith.Constant.from_int_and_width(1, builtin.i64),
+                res := llvm.AllocaOp.get(lit1,
+                                         builtin.IntegerType.from_width(
+                                             8 * self.info.status_size),
+                                         as_untyped_ptr=True),
+            ], [res.res], res
+
+    def _emit_memref_counts(
+            self, ssa_val: SSAValue) -> tuple[list[Operation], OpResult]:
+        """
+        This takes in an SSA Value holding a memref, and creates operations
+        to calculate the number of elements in the memref.
+
+        It then returns a list of operations calculating that size, and
+        an OpResult containing the calculated value.
+        """
+        assert isinstance(ssa_val.typ, memref.MemRefType)
+
+        # Note: we only allow MemRef, not UnrankedMemref!
+        # TODO: handle -1 in sizes
+        if not all(dim.value.data >= 0 for dim in ssa_val.typ.shape.data):
+            raise RuntimeError(
+                "MPI lowering does not support unknown-size memrefs!")
+
+        size = sum(dim.value.data for dim in ssa_val.typ.shape.data)
+
+        literal = arith.Constant.from_int_and_width(size, mpi.t_int)
+        return [literal], literal.result
+
+    def _emit_mpi_type_load(self, type_attr: Attribute) -> Operation:
+        """
+        This emits an instruction loading the correct magic MPI value for the
+        xDSL type of <type_attr> into an SSA Value.
+        """
+        return arith.Constant.from_int_and_width(
+            self._translate_to_mpi_type(type_attr), mpi.t_int)
+
+    def _translate_to_mpi_type(self, typ: Attribute) -> int:
+        """
+        This translates an xDSL type to a corresponding MPI type
+
+        Currently supported mappings are:
+            floats:
+                f32     -> MPI_FLOAT
+                f64     -> MPI_DOUBLR
```

Comment:
```suggestion
                f64     -> MPI_DOUBLE
```

Reply:
Could you please also take a look at this?